src/perftest_resources.c

```diff
@@ -1566,29 +1566,129 @@ int destroy_ctx(struct pingpong_context *ctx,
  *
  ******************************************************************************/
 #ifdef HAVE_EX_ODP
-static int check_odp_support(struct pingpong_context *ctx)
+static int check_odp_support(struct pingpong_context *ctx, struct perftest_parameters *user_param)
 {
 	struct ibv_device_attr_ex dattr;
+	VerbType verb = user_param->verb;
+	MachineType machine = user_param->machine;
+	int conn = user_param->connection_type;
 	int ret = ibv_query_device_ex(ctx->context, NULL, &dattr);
 
 	if (ret) {
-		fprintf(stderr, " Couldn't query device for on-demand paging capabilities.\n");
+		fprintf(stderr, " Couldn't query device for On-Demand Paging capabilities.\n");
 		return 0;
 	}
 
 	/* These capabilities must be set by device drivers. */
 	if (!(dattr.odp_caps.general_caps & IBV_ODP_SUPPORT)) {
 		fprintf(stderr, " The device does not support On-Demand Paging.\n");
-                return 0;
-        }
+		return 0;
+	}
+
+	switch (conn) {
+	case RC:
+		if (verb == SEND && machine == CLIENT) {
+			if (!(dattr.odp_caps.per_transport_caps.rc_odp_caps & IBV_ODP_SUPPORT_SEND)) {
+				fprintf(stderr, " ODP Send is not supported for RC transport.\n");
+				return 0;
+			}
+		} else if (verb == SEND && machine == SERVER) {
+			if (!(dattr.odp_caps.per_transport_caps.rc_odp_caps &
+			      (IBV_ODP_SUPPORT_RECV | IBV_ODP_SUPPORT_SRQ_RECV))) {
+				fprintf(stderr, " ODP Recv is not supported for RC transport.\n");
+				return 0;
+			}
+		} else if (verb == WRITE &&
+			   !(dattr.odp_caps.per_transport_caps.rc_odp_caps & IBV_ODP_SUPPORT_WRITE)) {
+			fprintf(stderr, " ODP Write is not supported for RC transport.\n");
+			return 0;
+		} else if (verb == READ &&
+			   !(dattr.odp_caps.per_transport_caps.rc_odp_caps & IBV_ODP_SUPPORT_READ)) {
+			fprintf(stderr, " ODP Read is not supported for RC transport.\n");
+			return 0;
+		} else if (verb == ATOMIC &&
+			   !(dattr.odp_caps.per_transport_caps.rc_odp_caps & IBV_ODP_SUPPORT_ATOMIC)) {
+			fprintf(stderr, " ODP Atomics are not supported for RC transport.\n");
+			return 0;
+		}
+		break;
 
-	if (!(dattr.odp_caps.per_transport_caps.rc_odp_caps & IBV_ODP_SUPPORT_SEND)) {
-		fprintf(stderr, " Send is not supported for RC transport.\n");
+	case UC:
+                fprintf(stderr," ODP is not available on UC transport.\n");
 		return 0;
-	} else if (!(dattr.odp_caps.per_transport_caps.rc_odp_caps & IBV_ODP_SUPPORT_RECV)) {
-		fprintf(stderr, " Receive is not supported for RC transport.\n");
+
+	case UD:
+		if (verb == SEND && machine == CLIENT) {
+			if (!(dattr.odp_caps.per_transport_caps.ud_odp_caps & IBV_ODP_SUPPORT_SEND)) {
+				fprintf(stderr, " ODP Send is not supported for UD transport.\n");
+				return 0;
+			}
```

Comment:
We see a lot of duplication code in this change, can we have it in better shape ? 
like create a func that can do this check and print generated string.

Reply:
Thanks for the comment! I modified the patch.